app/permissions_migrations.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import "github.com/mattermost/mattermost-server/model"
+
+type permissionTransformation struct {
+	On     func(map[string]bool) bool
+	Add    []string
+	Remove []string
+}
+type permissionsMap []permissionTransformation
+
+const (
+	MIGRATION_KEY_EMOJI_PERMISSIONS_SPLIT   = "emoji_permissions_split"
+	MIGRATION_KEY_WEBHOOK_PERMISSIONS_SPLIT = "webhook_permissions_split"
+
+	PERMISSION_MANAGE_EMOJIS                   = "manage_emojis"
+	PERMISSION_MANAGE_OTHERS_EMOJIS            = "manage_others_emojis"
+	PERMISSION_CREATE_EMOJIS                   = "create_emojis"
+	PERMISSION_DELETE_EMOJIS                   = "delete_emojis"
+	PERMISSION_DELETE_OTHERS_EMOJIS            = "delete_others_emojis"
+	PERMISSION_MANAGE_WEBHOOKS                 = "manage_webhooks"
+	PERMISSION_MANAGE_OTHERS_WEBHOOKS          = "manage_others_webhooks"
+	PERMISSION_MANAGE_INCOMING_WEBHOOKS        = "manage_incoming_webhooks"
+	PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS = "manage_others_incoming_webhooks"
+	PERMISSION_MANAGE_OUTGOING_WEBHOOKS        = "manage_outgoing_webhooks"
+	PERMISSION_MANAGE_OTHERS_OUTGOING_WEBHOOKS = "manage_others_outgoing_webhooks"
+)
+
+func permissionExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return ok && val
+	}
+}
+
+func permissionNotExists(permission string) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		val, ok := permissions[permission]
+		return !(ok && val)
+	}
+}
+
+func permissionOr(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
```

Comment:
I'm a little out of the loop on the greater context... I see that `permissionOr` and `permissionAnd` are not used yet... is their use eminent or is it possible they'll be added and never used?

Reply:
Would be very weird to not need them in the future, the idea is to have the posibility of create permissions, and give them a default value based on the existing ones, for example, if you create a new permission call "MOVE_EMOJI" maybe you want to set this for any user who have "ADD_REACTION" and "REMOVE_REACTION". There are a lot of valid cases where you can want derive your new permission based on current permissions.